shim/src/repair/functionConstructors.js

```diff
@@ -0,0 +1,68 @@
+// Adapted from SES/Caja - Copyright (C) 2011 Google Inc.
+// https://github.com/google/caja/blob/master/src/com/google/caja/ses/startSES.js
+// https://github.com/google/caja/blob/master/src/com/google/caja/ses/repairES5.js
+
+import verifyStrictFunctionBody from '../evaluate/verifyStrictFunctionBody';
+import { getPrototypeOf, setPrototypeOf, defineProperty } from '../commons';
+
+// This block replaces the original Function, %GeneratorFunction% and
+// %AsycnGeneratorFunction%, with safe implementations that preserve SES
+// confinement. After this block is done, the originals should no longer be
+// reachable.
+
+export default function repairFunctionConstructors(confinedWindow) {
+
+    const { eval: secureEval } = confinedWindow;
+
+    /**
+     * A safe form of the {@code Function} constructor, which
+     * constructs strict functions that can only refer freely to the
+     * {@code sharedImports}.
+     *
+     * The returned function is strict whether or not it declares
+     * itself to be.
+     */
+    function secureFunctionFactory(funcDeclaration) {
+        return (function SecureFunction(...params) {
+            var body = verifyStrictFunctionBody(params.pop() || '');
+
+            var exprSrc = `
+(${funcDeclaration}(${params.join(',')} {
+    ${body}
+})`;
+            return secureEval(exprSrc);
+        });
+    }
+
+    // Our process to secure constructors:
+    // 1. Obtain the original prototype of an instance from the sandbox
+    // 2. Create a secure constructor
+    // 3. Replace its prototype property with the original prototype
+    // 4*. Replace its internal [[Prototype]] property with a secure function (except for Function)
+    // 5. Replace the original prototype's constructor with the secure constructor
+    // 6*. Export the secure function to the sandbox (only for Function)
+
+    const FuncProto = getPrototypeOf(secureEval('(function () {})'));
+    const secureFunction = secureFunctionFactory('function');
```

Comment:
The code is correct. I could have used the alternate syntax:
```
Reflect.getPrototypeOf(eval('(function () {})')) === global.Function.prototype
// true
```

But using `Function` would not have been right. `Function` and `Function.prototype` are different. However, `Function` and `Function.constructor` are the same.

To avoid that trap and to ease maintenance, I wanted the code to be symmetric, since the "repair" process is complex but always the same. I repeatedly use the same technique for Function, FunctionGenerator and AsyncFunction.  

The only difference with Mark's original [code](https://github.com/tc39/proposal-realms/issues/54) is to support a given `sandbox`:
```
const UnsafeFunc = Function;
```
That code was not written to use a provided context, such as "sandbox", but was using the `Function` as visible in the current execution context.

Idem with FunctionGenerator which was also using the current execution context:
```
const Generator = Reflect.getPrototypeOf(function*(){});
```